Source/Core/VideoCommon/ShaderCache.cpp

```diff
@@ -627,6 +617,22 @@ static GXPipelineUid ApplyDriverBugs(const GXPipelineUid& in)
     ps->ztest = EmulatedZ::Early;
   }
 
+  // We can use framebuffer fetch to emulate logic ops in the fragment shader.
+  // So don't use the worse blend approximation.
+  if (blend.logicopenable && !g_ActiveConfig.backend_info.bSupportsLogicOp &&
+      !g_ActiveConfig.backend_info.bSupportsFramebufferFetch)
+  {
+    if (!blend.LogicOpApproximationIsExact())
+      WARN_LOG_FMT(VIDEO,
+                   "Approximating logic op with blending, this will produce incorrect rendering.");
+    if (blend.LogicOpApproximationWantsShaderHelp())
+    {
+      ps->emulate_logic_op_with_blend = true;
+      ps->logic_op_mode = static_cast<u32>(blend.logicmode.Value());
+    }
+    blend.ApproximateLogicOpWithBlending();
```

Comment:
My suggestion:

Leave `ApproximateLogicOpWithBlending()` alone, the name properly explains what it is doing.

It looks like `LogicOpApproximationIsExact()` is doing the same thing as the booleans you added to `approximations` in `ApproximateLogicOpWithBlending()`.  If it is, then I'd suggest to change this block to look like:

```
  if (blend.LogicOpApproximationWantsShaderHelp())
  {
      ps->emulate_logic_op_with_blend = true;
      ps->logic_op_mode = static_cast<u32>(blend.logicmode.Value());
  }
  if (!blend.LogicOpApproximationIsExact())
  {
    WARN_LOG_FMT(VIDEO, "Approximating...");
    blend.ApproximateLogicOpWithBlending();
  }
```

Reply:
They're different.  The new boolean in `approximations` is whether blending should be enabled.  `LogicOpApproximationIsExact` says whether an approximation gives identical results to the real thing.  As an example of a logic op where the two aren't the same, `Invert` has blending enabled but perfectly emulates the logic op, and therefore returns true from `LogicOpApproximationIsExact`.

Reply:
Ok, looking like the boolean's are slightly different.  I'd suggest adding a new function:

```
bool ShouldApproximateLogicOpWithBlending()
{
  switch (logicmode.Value())
  {
  case LogicOp::And:
  case LogicOp::AndReverse:
  case LogicOp::AndInverted:
  case LogicOp::NoOp:
  case LogicOp::Xor:
  case LogicOp::Or:
  case LogicOp::Nor:
  case LogicOp::Equiv:
  case LogicOp::Invert:
  case LogicOp::OrReverse:
  case LogicOp::OrInverted:
  case LogicOp::Nand:
    return true;
  default:
    return false;
  }
}
```

and use that instead:

```
  if (blend.LogicOpApproximationWantsShaderHelp())
  {
      ps->emulate_logic_op_with_blend = true;
      ps->logic_op_mode = static_cast<u32>(blend.logicmode.Value());
  }
  if (!blend.ShouldApproximateLogicOpWithBlending())
  {
    WARN_LOG_FMT(VIDEO, "Approximating logic op with blending, this will produce incorrect rendering.");
    blend.ApproximateLogicOpWithBlending();
  }
```

It might seem minor but with this change, you are telling the user exactly what you're doing and the function names match up too.